---
title: "Microbiome Analysis"
output: html_notebook
author: "Joseph Edwards"
---

## Introduction
In this R notebook, we will be conducting an example analysis of microbiome data from the roots of plants. This is not a comprehensive analysis, but the workflow here gives a barebones structure for a typical analysis of amplicon data. If the reader has any questions or concerns, please contact Joseph Edwards at edwards@ucdavis.edu or Chistian Santos at cmsantosm@ucdavis.edu.

## Background on the data
This data is derived from Santos-Medellín et al. [Drought stress results in a compartment-specific restructuring of the rice root-associated microbiomes. *MBio*, 2017](http://mbio.asm.org/content/8/4/e00764-17.short). In this paper, the authors inspect the microbiomes of various rice genotypes when exposed to drought in different soils. The OTU table for this particular tutorial was generated only from the well-watered plants in this study.

## Required libraries
Here we will be loading in the necessary libraries for the example analysis. The tidyverse contains the libraries ggplot2, tibble, tidyr, readr, purrr, dplyr, stringr, and forcats. The tidyMB package contains a bunch of useful wrapper functions to make microbiome analysis streamlined with a tidy workflow. The vegan package contains many tools essential for community (and microbial) ecology.
```{r}
library(tidyverse)
library(tidyMB)
library(vegan)
library(reshape2)
```

## Load in the data
Here we will load in the OTU table, the mapping file (metadata), and the taxonomy table. The OTU table and mapping file are tab delimited files generated freshly for this analysis. If the user would like to generate these files for themselves, please follow the instructions (here)[https://github.com/bulksoil/BioProtocol/blob/master/ANALYSIS/SeqProcessing/seq_processing.md]. The taxonomy table is a pre-compiled table that is in a ready-to-load R data structure. The OTU table was constructed using closed reference clustering. Therefore there was no need to assign taxonomies to the OTUs.
```{r}
otu_table <- read.table("../../DATA/example_otu_table.tsv", sep = "\t", header = T, row.names = 1)
map_table <- read_tsv("../../DATA/example_map.tsv")
tax_table <- read_rds("../../DATA/gg_otus_tax.rds")
```

## Filtering the data
Here is where the user will clean up the data. This includes removing plastid and mitochondria OTUs from the dataset, filtering out low-depth samples, and removing OTUs which are not regularly present in the samples. We will remove OTUs which are not present in at least 5% of the samples.

### Assemble data into a 'tidy' format
A good rule of thumb in any analysis is to convert data into a tidy format. To be tidy, the dataset should have one observation per row and the additional columns should reflect information about that observation.
```{r}
map_table <- map_table[match(colnames(otu_table), map_table$SampleID),]
mb_experiment <- melt(cbind(map_table, t(otu_table)), id.vars = names(map_table)) %>% 
  inner_join(tax_table, by = "variable")
```

### Distribution of sequencing depth
It's a good idea to first look at the distribution of sampling depth in each sample type. Sometimes samples are not well represented in the sequencing and should be removed before further analysis. 
```{r}
mb_experiment %>% 
  group_by(SampleID, Compartment) %>% 
  summarise(Depth = sum(value)) %>% 
  ggplot(aes(Depth, fill = Compartment)) +
  geom_histogram() +
  scale_fill_manual(values = c("#E41A1C", "#377EB8", "#984EA3")) +
  theme_minimal() +
  facet_grid(.~Compartment)
```
Here, it appears that all the samples are fine.

### Identify and remove plastid and mitochondria OTUs and define a column for sequencing depth
```{r}
mb_experiment <- mb_experiment %>% 
  filter(Family != "mitochondria") %>% 
  filter(Class != "Chloroplast") %>% 
  group_by(SampleID) %>% 
  mutate(Depth = sum(value))
```

### Remove OTUs which are not present in at least 5% of the samples
Let us first add a few columns to our data. Here we will add a column for the prevalence of each OTU, the relative abundance of each OTU (per mille), and a column specifying whether the column will be kept after filtering based upon a 5% prevalence threshold.
```{r}
mb_experiment <- mb_experiment %>% 
  group_by(variable) %>% 
  mutate(prev = sum(value > 0) / n()) %>% 
  mutate(RA = (value / Depth) * 1000) %>% 
  mutate(Status = ifelse(prev >= 0.05, "Keep", "Discard"))
```

Next we will make a graph plotting mean relative abundance vs. prevalence. The points represent each OTU and are colored by whether they will be kept after filtering.
```{r}
prev_plot <- mb_experiment %>% 
  group_by(variable, prev, Status) %>% 
  summarise(mean_ra = mean(RA)) %>% 
  ggplot(aes(x = mean_ra, y = prev, color = Status)) +
  geom_point(alpha = 0.1) +
  scale_x_log10() +
  scale_color_manual(values = c("red", "black")) +
  labs(x = "Mean Relative Abundance (log10)", y = "Prevalence") +
  theme_minimal() +
  theme(text = element_text(size = 12), legend.position = "none")

prev_plot
```

Finally we will remove all the low prevalence OTUs
```{r}
mb_experiment <- mb_experiment %>% 
  filter(Status == "Keep")
```


## Beta Diversity
One of the first things microbial ecologists do when they finally have their data cleaned up and in a managable format is to identify the obvious underlying patterns using principal coordinate analysis. Here we will use the convenient tidy_pcoa wrapper function from the tidyMB package to perform a PCoA using a Bray-Curtis dissimilary metric between the samples.
```{r}
exp_pc <- tidy_pcoa(mb_experiment %>% 
                      mutate(logRA = log2(RA + 1)) %>% 
                      select(SampleID, Tub, Compartment, Soil, Genotype, variable, RA, logRA), 
                    value = "logRA", dist = "bray")
```

Plot it out
```{r}
pc_plot <- exp_pc$axes %>% 
  ggplot(aes(MDS1, MDS2, color = Compartment)) +
  geom_point(size = 3, alpha = 0.8) +
  labs(x = paste("PCo1 (", round(exp_pc$eigen_vals[1], 3)*100, "%)", sep = ""),
       y = paste("PCo2 (", round(exp_pc$eigen_vals[2], 3)*100, "%)", sep = "")) +
  theme_minimal() +
  scale_color_manual(values = c("#E41A1C", "#377EB8", "#984EA3")) +
  theme(text = element_text(size = 12), legend.position = "none")

pc_plot
```
From the results, we can see that the bulk soil and rhizosphere have very similar communities while the endosphere is very different from both of these compartments. Further analysis will tell us which phyla and OTUs differ between the compartments and how diverse each community is.

## Alpha Diversity
Alpha diversity is a measure of how diverse any given sample is (i.e. how many microbes are in the sample or how even are their distributions?). There are a few different measures of alpha diversity and the results can be different depending on the metric. 

Here, we will produce two measures of alpha diversity - species richness and Shannon's entropy. Species richness is simply counting how many different units are detected within a sample. Shannon's entropy requires that the abundances be in a percentage (rather than per mille) form.
```{r}
mb_a_div <- mb_experiment %>% 
  group_by(SampleID, Compartment, Soil) %>% 
  mutate(RA = RA /1000) %>% 
  filter(value > 0) %>% 
  summarise(Richness = sum(value > 0), Shannon = -sum(log(RA) * RA))
```

Plot it out.
```{r}
adiv_plot <- mb_a_div %>% 
  ungroup() %>%
  mutate(Compartment = fct_relevel(Compartment, "BulkSoil", "Rhizosphere", "Endosphere")) %>% 
  gather(metric, value, -c(SampleID, Compartment, Soil)) %>% 
  ggplot(aes(Compartment, value, fill = Compartment)) +
  geom_boxplot() +
  facet_wrap(~metric, scales = "free") +
  theme_minimal() +
  labs(x = "", y = "Value") +
  scale_fill_manual(values = c("#E41A1C", "#984EA3", "#377EB8")) +
  theme(text = element_text(size = 12), 
        legend.position = "none", 
        axis.text.x = element_text(angle = 30, hjust = 1))

adiv_plot
```
Firstly, notice that the graphs have different scales on the y-axis but overall, the patterns look very similar. Species richness and Shannon entropy are measureing different things, but the conclusions drawn from these two techniques can be similar. Here we notice that the bulk soil and the rhizosphere have similar levels of diversity while the endosphere has a reduced diversity compared to the other two sample types. We infer this to mean that the endosphere is a niche that fewer of the microbes in the soil can reside in compared to the rhizosphere.

## Phylum Plots
Perhaps the researcher wants to know the microbial patterns at a higher level than OTU. By looking at phylum level abundances, one can get a sense of how large the differences in community structures are between any given comparison.

Each OTU in the dataset has taxonomic information assigned to it, so to begin, we will aggregate the counts for each phylum in each sample. Next, the chunk of code nests the data and sums up the total abundance for each detected phylum across the dataset. We filter the 10 most highly represent phyla and take a mean of their relative abundance across sample types.
```{r}
phy_abundance <- mb_experiment %>% 
  group_by(SampleID, Phylum, SampleID, Compartment, Soil) %>% 
  summarise(phy_total = sum(RA)) %>% 
  group_by(Phylum) %>% 
  nest() %>% 
  mutate(total = map_dbl(data, ~sum(.x$phy_total))) %>% 
  top_n(10, total) %>% 
  select(-total) %>% 
  unnest(data) %>% 
  group_by(Phylum, Compartment, Soil) %>% 
  summarise(mean_phy = mean(phy_total)) %>% 
  ungroup() %>% 
  mutate(Compartment = fct_relevel(Compartment, "BulkSoil", "Rhizosphere", "Endosphere"))
```

Plot it out
```{r}
phy_plot <- phy_abundance %>% 
  ggplot(aes(x = Compartment, mean_phy, fill = Phylum)) +
  geom_bar(stat = "identity") +
  scale_fill_brewer(palette = "Spectral") +
  facet_grid(.~Soil) +
  theme_minimal() +
  labs(x = "", y = "Relative Abundance (‰)") +
  theme(axis.text.x = element_text(angle = 30, hjust = 1), text = element_text(size = 12))

phy_plot
```
Here we can see that there appear to be statistically different distributions of phyla in each compartment. In each tested soil the bulk soil and rhizopshere have a lower relative abundance of proteobacteria compared the endosphere, while the opposite distribution appears for Acidobacteria.


## Differential abundance analysis
Here we will compare the microbiomes of the rhizosphere and endosphere to that of the bulk soil to see which OTUs are differentially abundant and describing the witnessed differences between the sample types. For this particular example we will be using DESeq2 to identify the OTUs which prefer one compartment compared to another. Again, this will be done using a tidy framework in order to prevent the creation of excessive dataframes. Essential to this approach are the tidyMB and biobroom packages. 

In the chunk of code below, we filter for OTUs which are represented in 10% or more samples and next load the data into an object useful for the DESeq2 package (using the tidyDGEL function). Because we are interested in understanding which OTUs are differentially abunant between the compartments, our grouping variable is the Compartment column. The DESeq command will then run all of the processes in the DESeq2 workflow. Finally, lfcShrink identify the OTUs differentially abundant between the levels provided in the contrasts. 
```{r}
library(DESeq2)
library(broom)
library(biobroom)
DESeq2_example <- mb_experiment %>% 
  select(Compartment, variable, value, SampleID) %>% 
  group_by(variable) %>% 
  filter(sum(value > 0) / n() > 0.1) %>% 
  mutate(Compartment = relevel(factor(Compartment), ref = "BulkSoil")) %>% 
  mutate(group = Compartment) %>% 
  group_by(1) %>% 
  nest() %>% 
  mutate(DGEL = map(data, ~tidyDGEL(., value = "value", group_column = "group", method = "DESeq2", formula = "~ group"))) %>% 
  mutate(dds = map(DGEL, ~DESeq(.))) %>% 
  mutate(BulkSoil_Endosphere = map(dds, ~lfcShrink(., contrast = c("group", "BulkSoil", "Endosphere")))) %>% 
  mutate(BulkSoil_Rhizosphere = map(dds, ~lfcShrink(., contrast = c("group", "BulkSoil", "Rhizosphere")))) %>% 
  dplyr::select(BulkSoil_Endosphere, BulkSoil_Rhizosphere) %>% 
  gather(key = Comparison, value = results)
```

Because the resulting object is nested, we will go ahead and unnest it to look at the results. We can then classify those OTUs which are differentially abundant in some scenario (adjusted P value <= 0.05) as being enriched in one compartment compared to another.
```{r}
DESeq2_example_results <- DESeq2_example %>% 
  unnest(map(results, ~tidy(.))) %>% 
  tidyr::separate(Comparison, c("Comp1", "Comp2"), "_", remove = F) %>% 
  mutate(direction = ifelse(estimate > 0, Comp1, Comp2)) %>% 
  mutate(classification = ifelse(p.adjusted <= 0.05, direction, "ns")) %>% 
  replace_na(list(classification = "ns")) %>% 
  mutate(Comparison = gsub("Soil_", " Soil vs. ", Comparison)) %>% 
  mutate(Comparison = fct_relevel(Comparison, "Bulk Soil vs. Rhizosphere", "Bulk Soil vs. Endosphere"))
```

There are many things that one can do with this dataset. One of the most common ways of visualizing patterns is to make a MA plot with the mean abundance on the x-axis and the log-fold change between the two compartments on the y-axis. Each point represents one OTU and its color represents whethere the OTU was differentially abundant.
```{r}
MA_plot <- DESeq2_example_results %>% 
  ggplot(aes(baseMean, estimate, color = classification, alpha = classification)) +
  geom_point(size = 1) +
  facet_grid(.~Comparison) +
  scale_x_log10() +
  scale_color_manual(values = c("red", "steelblue", "grey30", "#984EA3")) +
  scale_alpha_manual(values = c(0.8, 0.8, 0.1, 0.8)) +
  labs(x = "Mean abundance (log10)", y = "Log Fold Change") + 
  theme_minimal() +
  theme(text = element_text(size = 12), legend.position = "none")

MA_plot
```
In this plot we can see that very few of the microbes in the bulk soil are depleted from the rhizosphere. It appears that rhizosphere has an enrichment role rather than a depletion role. In the endosphere, we see many OTUs which are enriched compared to the bulk soil and many bulk soil OTUs which are depleted. This makes sense as the endosphere is a nutrient rich niche that has many physical and host-associated barriers for colonization.

## Conclusion
Here we have presented a small introduction to plant root microbiome data analysis generated from 16S rRNA gene sequencing. This is by no means a comprehensive workflow, but it can serve as a good introduction to microbiome data analysis for a user with little experience. We have emphasized the use of tidy tools in order to make the code more readable (and hopefully easier to reproduce). We hope that the reader has come away with several points.

1. Do not reinvent the wheel
    + Use the packages that already exist in order to maximize reproducibilty and minimize mistakes
2. Drawn on approaches that have been used across disciplines
    + DESeq2 was not designed for microbiome studies, but it works quite well microbiome purposes. Other approaches designed for other fields can also work for microbiome analysis.
3. Use a tidy workflow
    + This will ensure that the code is compact, cleaner, and easier to read than alternative methods.



